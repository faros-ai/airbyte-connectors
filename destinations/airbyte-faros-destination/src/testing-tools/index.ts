export * from './cli';
export * from './testing-tools';
export * from './utils';
