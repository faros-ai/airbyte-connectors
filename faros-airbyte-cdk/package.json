{
  "name": "faros-airbyte-cdk",
  "version": "0.0.1",
  "description": "Airbyte Connector Development Kit (CDK) for JavaScript/TypeScript",
  "keywords": [
    "airbyte",
    "cdk",
    "faros"
  ],
  "homepage": "https://www.faros.ai",
  "author": "Faros AI, Inc.",
  "license": "Apache-2.0",
  "repository": "https://github.com/faros-ai/airbyte-connectors",
  "bugs": {
    "url": "https://github.com/faros-ai/airbyte-connectors/issues"
  },
  "directories": {
    "lib": "lib",
    "test": "test"
  },
  "files": [
    "lib/",
    "resources/"
  ],
  "main": "./lib",
  "engines": {
    "node": ">=16"
  },
  "scripts": {
    "build": "tsc -p src",
    "clean": "rm -rf lib node_modules out",
    "fix": "prettier --write 'src/**/*.ts' 'test/**/*.ts' && npm run lint -- --fix",
    "lint": "eslint 'src/**/*.ts' 'test/**/*.ts'",
    "prepare": "npm run build",
    "test": "jest --verbose --color",
    "test-cov": "jest --coverage --verbose --color",
    "watch": "tsc -b -w src test"
  },
  "dependencies": {
    "@apidevtools/json-schema-ref-parser": "^9.0.9",
    "commander": "^9.3.0",
    "enquirer": "^2.3.6",
    "fast-redact": "^3.0.2",
    "json-schema-traverse": "^1.0.0",
    "lodash": "^4.17.21",
    "pino": "^8.5.0",
    "table": "^6.8.1",
    "verror": "^1.10.1"
  },
  "jest": {
    "coverageDirectory": "out/coverage",
    "coveragePathIgnorePatterns": [
      "<rootDir>/node_modules/",
      "<rootDir>/test/"
    ],
    "preset": "ts-jest",
    "testEnvironment": "node",
    "testTimeout": 10000,
    "globals": {
      "ts-jest": {
        "tsconfig": "test/tsconfig.json"
      }
    }
  }
}
