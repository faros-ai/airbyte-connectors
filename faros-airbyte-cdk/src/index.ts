export * from './destinations/destination';
export * from './destinations/destination-runner';
export {wrapApiError} from './errors';
export * from './help';
export * from './logger';
export * from './protocol';
export * from './sources/source-base';
export * from './sources/source';
export * from './sources/source-runner';
export * from './sources/streams/stream-base';
export * from './spec-loader';
export * from './utils';
