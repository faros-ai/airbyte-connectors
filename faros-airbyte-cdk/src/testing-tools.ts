import fs from 'fs';
import {Dictionary} from 'ts-essentials';

import {
  AirbyteConfig,
  AirbyteConfiguredCatalog,
  AirbyteMessageType,
  AirbyteRecord,
  AirbyteSourceBase,
  AirbyteState,
} from '.';

export function readTestResourceFile(fileName: string): string {
  return fs.readFileSync(`test/resources/${fileName}`, 'utf8');
}

/**
 * Parse a test resource into JSON
 */
export function readTestResourceAsJSON(fileName: string): any {
  return JSON.parse(readTestResourceFile(fileName));
}

export interface SourceCheckTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath?: string | AirbyteConfig;
}

export const sourceCheckTest = async (
  options: SourceCheckTestOptions
): Promise<void> => {
  const {source, configOrPath} = options;
  const config = resolveInput(configOrPath);
  expect(await source.check(config)).toMatchSnapshot();
};

export interface SourceReadTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath: string | AirbyteConfig;
  catalogOrPath: string | AirbyteConfiguredCatalog;
  stateOrPath?: string | AirbyteState;
  onBeforeReadResultConsumer?: (res: {
    config: AirbyteConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }) => void;
  checkRecordsData?: (records: ReadonlyArray<Dictionary<any>>) => void;
}

export const sourceReadTest = async (
  options: SourceReadTestOptions
): Promise<void> => {
  const {
    source,
    configOrPath,
    catalogOrPath,
    stateOrPath = undefined,
    onBeforeReadResultConsumer = undefined,
    checkRecordsData = undefined,
  } = options;
  const config = resolveInput(configOrPath);
  const catalog = resolveInput(catalogOrPath);
  const state = stateOrPath ? resolveInput(stateOrPath) : {};
  const res = await source.onBeforeRead(config, catalog, state);
  if (onBeforeReadResultConsumer) {
    onBeforeReadResultConsumer(res);
  }
  const generator = source.read(res.config, res.config, res.catalog, res.state);
  const records: Dictionary<any>[] = [];
  for await (const message of generator) {
    if (message.type === AirbyteMessageType.RECORD) {
      records.push((message as AirbyteRecord).record.data);
    }
  }
  if (checkRecordsData) {
    checkRecordsData(records);
  }
};

function resolveInput<T>(inputOrPath: string | T): T {
  return typeof inputOrPath === 'string'
    ? readTestResourceAsJSON(inputOrPath)
    : inputOrPath;
}
