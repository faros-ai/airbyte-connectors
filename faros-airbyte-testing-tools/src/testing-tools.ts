import fs from 'fs';
import {Mockttp} from 'mockttp';
import pino from 'pino';
import tmp from 'tmp-promise';
import {Dictionary} from 'ts-essentials';
import util from 'util';

import {
  AirbyteConfig,
  AirbyteConfiguredCatalog,
  AirbyteMessageType,
  AirbyteRecord,
  AirbyteSourceBase,
  AirbyteState,
  AirbyteStateMessage,
} from 'faros-airbyte-cdk';
import {Data} from 'faros-airbyte-cdk/lib/utils';
import {Edition, InvalidRecordStrategy} from 'airbyte-faros-destination/lib/common/types';

const TEST_SOURCE_ID = 'mytestsource';

tmp.setGracefulCleanup();

/**
 * Read a test resource by name
 */
export function readTestResourceFile(fileName: string): string {
  return fs.readFileSync(`test/resources/${fileName}`, 'utf8');
}

const writeFile = util.promisify(fs.write);

/**
 * Parse a test resource into JSON
 */
export function readTestResourceAsJSON(fileName: string): any {
  return JSON.parse(readTestResourceFile(fileName));
}

/**
 * Creates a temporary file
 * @return path to the temporary file
 */
export async function tempFile(data: string, postfix: string): Promise<string> {
  const file = await tmp.file({postfix});
  await writeFile(file.fd, data, null, 'utf-8');
  return file.path;
}

export function readTestFile(fileName: string): string {
  return fs.readFileSync(`test_files/${fileName}`, 'utf8');
}

/**
 * Parse a test file into JSON
 */
export function readTestFileAsJSON(fileName: string): any {
  return JSON.parse(readTestFile(fileName));
}

export function readResourceFile(fileName: string): string {
  return fs.readFileSync(`resources/${fileName}`, 'utf8');
}

/**
 * Parse a resource into JSON
 */
export function readResourceAsJSON(fileName: string): any {
  return JSON.parse(readResourceFile(fileName));
}

export interface TempConfigOptions {
  readonly api_url: string;
  readonly invalid_record_strategy?: InvalidRecordStrategy;
  readonly edition?: Edition;
  readonly edition_configs?: Dictionary<any>;
  readonly source_specific_configs?: Dictionary<any>;
  readonly replace_origin_map?: Dictionary<any>;
  readonly exclude_fields_map?: Dictionary<any>;
  readonly log_records?: boolean;
}

/**
 * Creates a temporary file with testing configuration
 * @return path to the temporary config file to delete
 */
export async function tempConfig(options: TempConfigOptions): Promise<string> {
  const conf = getConf(options);

  return tempFile(JSON.stringify(conf), '.json');
}

export function getConf(options: TempConfigOptions): any {
  const {
    api_url,
    invalid_record_strategy = InvalidRecordStrategy.FAIL,
    edition = Edition.CLOUD,
    edition_configs = {},
    source_specific_configs = {},
    replace_origin_map = {},
    exclude_fields_map = {},
    log_records = false,
  } = options;

  const edition_configs_defaults =
    edition === Edition.CLOUD
      ? {
          edition,
          api_url,
          api_key: 'test-api-key',
          graph: 'test-graph',
        }
      : {
          edition,
          hasura_url: api_url,
          segment_user_id: 'bacaf6e6-41d8-4102-a3a4-5d28100e642f',
          segment_test_host: api_url,
        };

  const conf = {
    edition_configs: {...edition_configs_defaults, ...edition_configs},
    invalid_record_strategy,
    origin: 'test-origin',
    jsonata_destination_models: ['generic_Record'],
    jsonata_expression: `
    data.{
      "model": "generic_Record",
      "record": {
        "uid": foo
      }
    }`,
    source_specific_configs,
    replace_origin_map: JSON.stringify(replace_origin_map),
    exclude_fields_map: JSON.stringify(exclude_fields_map),
    faros_source_id: TEST_SOURCE_ID,
    log_records,
  };

  return conf;
}

export function sourceSpecificTempConfig(
  url: string,
  source_specific_configs: Dictionary<any>
): Promise<string> {
  return tempConfig({api_url: url, source_specific_configs});
}

export async function initMockttp(mockttp: Mockttp): Promise<void> {
  await mockttp.start({startPort: 30000, endPort: 50000});

  await mockttp
    .forGet('/users/me')
    .once()
    .thenReply(200, JSON.stringify({tenantId: '1'}));

  await mockttp
    .forGet('/graphs/test-graph/statistics')
    .once()
    .thenReply(200, JSON.stringify({}));

  await mockttp.forGet('/healthz').once().thenReply(200, JSON.stringify({}));

  await mockttp
    .forGet(`/accounts/${TEST_SOURCE_ID}`)
    .once()
    .thenReply(200, JSON.stringify({account: {accountId: TEST_SOURCE_ID}}));

  const mockSyncResult = {
    sync: {
      syncId: '1',
      logId: '1',
      startedAt: new Date().toISOString(),
      status: 'running',
    },
  };
  await mockttp
    .forPut(`/accounts/${TEST_SOURCE_ID}/syncs`)
    .once()
    .thenReply(200, JSON.stringify(mockSyncResult));

  await mockttp
    .forPatch(`/accounts/${TEST_SOURCE_ID}/syncs/1`)
    .once()
    .thenReply(200, JSON.stringify(mockSyncResult));
}

export function testLogger(name = 'test'): pino.Logger {
  return pino({
    name,
    level: process.env.LOG_LEVEL ?? 'info',
    transport:
      process.env.LOG_LEVEL?.toLowerCase() === 'debug'
        ? {target: 'pino-pretty', options: {levelFirst: true}}
        : undefined,
  });
}

export interface SourceCheckTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath?: string | AirbyteConfig;
}

export const sourceCheckTest = async (
  options: SourceCheckTestOptions
): Promise<void> => {
  const {source, configOrPath} = options;
  const config = resolveInput(configOrPath);
  expect(await source.check(config)).toMatchSnapshot();
};

export interface SourceReadTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath: string | AirbyteConfig;
  catalogOrPath: string | AirbyteConfiguredCatalog;
  stateOrPath?: string | AirbyteState;
  onBeforeReadResultConsumer?: (res: {
    config: AirbyteConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }) => void;
  checkRecordsData?: (records: ReadonlyArray<Dictionary<any>>) => void;
  checkFinalState?: (state: Dictionary<any>) => void;
}

export const sourceReadTest = async (
  options: SourceReadTestOptions
): Promise<void> => {
  const {
    source,
    configOrPath,
    catalogOrPath,
    stateOrPath = undefined,
    onBeforeReadResultConsumer = undefined,
    checkRecordsData = undefined,
    checkFinalState = undefined,
  } = options;
  const config = resolveInput(configOrPath);
  const catalog = resolveInput(catalogOrPath);
  const state = stateOrPath ? resolveInput(stateOrPath) : {};
  const res = await source.onBeforeRead(config, catalog, state);
  if (onBeforeReadResultConsumer) {
    onBeforeReadResultConsumer(res);
  }
  const generator = source.read(res.config, res.config, res.catalog, res.state);
  const records: Dictionary<any>[] = [];
  let finalState: Dictionary<any>;
  for await (const message of generator) {
    if (message.type === AirbyteMessageType.RECORD) {
      records.push((message as AirbyteRecord).record.data);
    } else if (message.type === AirbyteMessageType.STATE) {
      finalState = (message as AirbyteStateMessage).state.data;
    }
  }
  if (checkRecordsData) {
    checkRecordsData(records);
  }
  if (checkFinalState) {
    checkFinalState(Data.decompress(finalState));
  }
};

export const sourceSchemaTest = (
  source: AirbyteSourceBase<AirbyteConfig>,
  config: AirbyteConfig
): void => {
  const streams = source.streams(config);

  const validateFieldInSchema = (
    field: string | string[],
    schema: Dictionary<any>
  ): void => {
    if (Array.isArray(field)) {
      let nestedField = schema;
      for (const subField of field) {
        expect(nestedField).toHaveProperty(subField);
        nestedField = nestedField[subField].properties;
      }
    } else {
      expect(schema).toHaveProperty(field);
    }
  };

  for (const stream of streams) {
    const jsonSchema = stream.getJsonSchema().properties;
    const primaryKey = stream.primaryKey;
    const cursorField = stream.cursorField;

    if (primaryKey) {
      if (Array.isArray(primaryKey)) {
        for (const key of primaryKey) {
          validateFieldInSchema(key, jsonSchema);
        }
      } else {
        validateFieldInSchema(primaryKey, jsonSchema);
      }
    }

    validateFieldInSchema(cursorField, jsonSchema);
  }
};

function resolveInput<T>(inputOrPath: string | T): T {
  return typeof inputOrPath === 'string'
    ? readTestResourceAsJSON(inputOrPath)
    : inputOrPath;
}

export const customStreamsTest = async (
  source: AirbyteSourceBase<AirbyteConfig>,
  config: AirbyteConfig,
  allStreams: ReadonlyArray<string>,
  selectedStreams?: ReadonlyArray<string>
): Promise<void> => {
  const {catalog} = await source.onBeforeRead(
    {...config, run_mode: 'Custom', custom_streams: selectedStreams},
    {streams: allStreams.map((name) => ({stream: {name}}))} as any
  );
  expect(catalog.streams).toHaveLength((selectedStreams ?? allStreams).length);
  expect(catalog.streams.map((s) => s.stream.name)).toEqual(
    expect.arrayContaining(selectedStreams ?? allStreams)
  );
};
