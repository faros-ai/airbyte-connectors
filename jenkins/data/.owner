2049311132 contextPath="" at 8@5674ab7e8d6d