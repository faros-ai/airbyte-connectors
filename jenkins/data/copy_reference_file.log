--- Copying files at Mon Sep  6 16:20:12 UTC 2021
--- Copying files at Mon Sep  6 19:51:19 UTC 2021
--- Copying files at Mon Sep  6 20:01:02 UTC 2021
