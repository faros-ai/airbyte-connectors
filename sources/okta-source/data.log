{"record":{"stream":"users","emitted_at":1641627744274,"data":{"id":"00u3jaazzxOxur0Bp5d7","status":"STAGED","created":"2022-01-08T06:50:58.000Z","activated":null,"statusChanged":null,"lastLogin":null,"lastUpdated":"2022-01-08T06:50:58.000Z","passwordChanged":null,"type":{"id":"oty3iw1gxxYagRUHa5d7"},"profile":{"firstName":"yen","lastName":"hoang","mobilePhone":null,"secondEmail":null,"login":"yenhh@gmail.com","email":"yenhh@gmail.com"},"credentials":{"emails":[{"value":"yenhh@gmail.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://dev-88164258.okta.com/api/v1/users/00u3jaazzxOxur0Bp5d7"}},"groupsOfUser":[{"id":"00g3iw1gxoucJXLqk5d7","created":"2022-01-07T06:09:36.000Z","lastUpdated":"2022-01-07T06:09:36.000Z","lastMembershipUpdated":"2022-01-08T06:50:58.000Z","objectClass":["okta:user_group"],"type":"BUILT_IN","profile":{"name":"Everyone","description":"All users in your organization"},"_links":{"logo":[{"name":"medium","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-medium.1a5ebe44c4244fb796c235d86b47e3bb.png","type":"image/png"},{"name":"large","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-large.d9cfbd8a00a4feac1aa5612ba02e99c0.png","type":"image/png"}],"users":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iw1gxoucJXLqk5d7/users"},"apps":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iw1gxoucJXLqk5d7/apps"}}},{"id":"00g3iwkxv69Dt4w3d5d7","created":"2022-01-07T07:53:52.000Z","lastUpdated":"2022-01-08T06:51:39.000Z","lastMembershipUpdated":"2022-01-08T06:50:58.000Z","objectClass":["okta:user_group"],"type":"OKTA_GROUP","profile":{"name":"test group","description":"group description"},"_links":{"logo":[{"name":"medium","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-medium.1a5ebe44c4244fb796c235d86b47e3bb.png","type":"image/png"},{"name":"large","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-large.d9cfbd8a00a4feac1aa5612ba02e99c0.png","type":"image/png"}],"users":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iwkxv69Dt4w3d5d7/users"},"apps":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iwkxv69Dt4w3d5d7/apps"}}}]}},"type":"RECORD"}
{"record":{"stream":"users","emitted_at":1641627745523,"data":{"id":"00u3iw1h2t4oi09Cv5d7","status":"ACTIVE","created":"2022-01-07T06:09:40.000Z","activated":null,"statusChanged":"2022-01-07T07:43:31.000Z","lastLogin":"2022-01-08T07:32:08.000Z","lastUpdated":"2022-01-07T07:43:31.000Z","passwordChanged":"2022-01-07T07:43:31.000Z","type":{"id":"oty3iw1gxxYagRUHa5d7"},"profile":{"firstName":"Kelvin","lastName":"Tran","mobilePhone":null,"secondEmail":null,"login":"kelvin.tr@mydevcube.com","email":"kelvin.tr@mydevcube.com"},"credentials":{"password":{},"emails":[{"value":"kelvin.tr@mydevcube.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://dev-88164258.okta.com/api/v1/users/00u3iw1h2t4oi09Cv5d7"}},"groupsOfUser":[{"id":"00g3iw1gxoucJXLqk5d7","created":"2022-01-07T06:09:36.000Z","lastUpdated":"2022-01-07T06:09:36.000Z","lastMembershipUpdated":"2022-01-08T06:50:58.000Z","objectClass":["okta:user_group"],"type":"BUILT_IN","profile":{"name":"Everyone","description":"All users in your organization"},"_links":{"logo":[{"name":"medium","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-medium.1a5ebe44c4244fb796c235d86b47e3bb.png","type":"image/png"},{"name":"large","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-large.d9cfbd8a00a4feac1aa5612ba02e99c0.png","type":"image/png"}],"users":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iw1gxoucJXLqk5d7/users"},"apps":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iw1gxoucJXLqk5d7/apps"}}},{"id":"00g3iwkxv69Dt4w3d5d7","created":"2022-01-07T07:53:52.000Z","lastUpdated":"2022-01-08T06:51:39.000Z","lastMembershipUpdated":"2022-01-08T06:50:58.000Z","objectClass":["okta:user_group"],"type":"OKTA_GROUP","profile":{"name":"test group","description":"group description"},"_links":{"logo":[{"name":"medium","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-medium.1a5ebe44c4244fb796c235d86b47e3bb.png","type":"image/png"},{"name":"large","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-large.d9cfbd8a00a4feac1aa5612ba02e99c0.png","type":"image/png"}],"users":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iwkxv69Dt4w3d5d7/users"},"apps":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iwkxv69Dt4w3d5d7/apps"}}}]}},"type":"RECORD"} 
{"record":{"stream":"groups","emitted_at":1641627748425,"data":{"id":"00g3iw1gxoucJXLqk5d7","created":"2022-01-07T06:09:36.000Z","lastUpdated":"2022-01-07T06:09:36.000Z","lastMembershipUpdated":"2022-01-08T06:50:58.000Z","objectClass":["okta:user_group"],"type":"BUILT_IN","profile":{"name":"Everyone","description":"All users in your organization"},"_links":{"logo":[{"name":"medium","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-medium.1a5ebe44c4244fb796c235d86b47e3bb.png","type":"image/png"},{"name":"large","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-large.d9cfbd8a00a4feac1aa5612ba02e99c0.png","type":"image/png"}],"users":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iw1gxoucJXLqk5d7/users"},"apps":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iw1gxoucJXLqk5d7/apps"}},"usersOfGroup":[{"id":"00u3iw1h2t4oi09Cv5d7","status":"ACTIVE","created":"2022-01-07T06:09:40.000Z","activated":null,"statusChanged":"2022-01-07T07:43:31.000Z","lastLogin":"2022-01-08T07:32:08.000Z","lastUpdated":"2022-01-07T07:43:31.000Z","passwordChanged":"2022-01-07T07:43:31.000Z","type":{"id":"oty3iw1gxxYagRUHa5d7"},"profile":{"firstName":"Kelvin","lastName":"Tran","mobilePhone":null,"secondEmail":null,"login":"kelvin.tr@mydevcube.com","email":"kelvin.tr@mydevcube.com"},"credentials":{"password":{},"emails":[{"value":"kelvin.tr@mydevcube.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://dev-88164258.okta.com/api/v1/users/00u3iw1h2t4oi09Cv5d7"}}},{"id":"00u3jaazzxOxur0Bp5d7","status":"STAGED","created":"2022-01-08T06:50:58.000Z","activated":null,"statusChanged":null,"lastLogin":null,"lastUpdated":"2022-01-08T06:50:58.000Z","passwordChanged":null,"type":{"id":"oty3iw1gxxYagRUHa5d7"},"profile":{"firstName":"yen","lastName":"hoang","mobilePhone":null,"secondEmail":null,"login":"yenhh@gmail.com","email":"yenhh@gmail.com"},"credentials":{"emails":[{"value":"yenhh@gmail.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://dev-88164258.okta.com/api/v1/users/00u3jaazzxOxur0Bp5d7"}}}]}},"type":"RECORD"}
{"record":{"stream":"groups","emitted_at":1641627749621,"data":{"id":"00g3iwkxv69Dt4w3d5d7","created":"2022-01-07T07:53:52.000Z","lastUpdated":"2022-01-08T06:51:39.000Z","lastMembershipUpdated":"2022-01-08T06:50:58.000Z","objectClass":["okta:user_group"],"type":"OKTA_GROUP","profile":{"name":"test group","description":"group description"},"_links":{"logo":[{"name":"medium","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-medium.1a5ebe44c4244fb796c235d86b47e3bb.png","type":"image/png"},{"name":"large","href":"https://ok12static.oktacdn.com/assets/img/logos/groups/odyssey/okta-large.d9cfbd8a00a4feac1aa5612ba02e99c0.png","type":"image/png"}],"users":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iwkxv69Dt4w3d5d7/users"},"apps":{"href":"https://dev-88164258.okta.com/api/v1/groups/00g3iwkxv69Dt4w3d5d7/apps"}},"usersOfGroup":[{"id":"00u3iw1h2t4oi09Cv5d7","status":"ACTIVE","created":"2022-01-07T06:09:40.000Z","activated":null,"statusChanged":"2022-01-07T07:43:31.000Z","lastLogin":"2022-01-08T07:32:08.000Z","lastUpdated":"2022-01-07T07:43:31.000Z","passwordChanged":"2022-01-07T07:43:31.000Z","type":{"id":"oty3iw1gxxYagRUHa5d7"},"profile":{"firstName":"Kelvin","lastName":"Tran","mobilePhone":null,"secondEmail":null,"login":"kelvin.tr@mydevcube.com","email":"kelvin.tr@mydevcube.com"},"credentials":{"password":{},"emails":[{"value":"kelvin.tr@mydevcube.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://dev-88164258.okta.com/api/v1/users/00u3iw1h2t4oi09Cv5d7"}}},{"id":"00u3jaazzxOxur0Bp5d7","status":"STAGED","created":"2022-01-08T06:50:58.000Z","activated":null,"statusChanged":null,"lastLogin":null,"lastUpdated":"2022-01-08T06:50:58.000Z","passwordChanged":null,"type":{"id":"oty3iw1gxxYagRUHa5d7"},"profile":{"firstName":"yen","lastName":"hoang","mobilePhone":null,"secondEmail":null,"login":"yenhh@gmail.com","email":"yenhh@gmail.com"},"credentials":{"emails":[{"value":"yenhh@gmail.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://dev-88164258.okta.com/api/v1/users/00u3jaazzxOxur0Bp5d7"}}}]}},"type":"RECORD"}
