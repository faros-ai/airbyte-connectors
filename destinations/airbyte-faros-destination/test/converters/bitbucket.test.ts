import {AirbyteLog, AirbyteLogLevel} from 'faros-airbyte-cdk';
import _ from 'lodash';
import {getLocal} from 'mockttp';

import {CLI, read} from '../cli';
import {initMockttp, tempConfig, testLogger} from '../testing-tools';
import {bitbucketAllStreamsLog} from './data';

describe('bitbucket', () => {
  const logger = testLogger();
  const mockttp = getLocal({debug: false, recordTraffic: false});
  const catalogPath = 'test/resources/bitbucket/catalog.json';
  let configPath: string;
  const streamNamePrefix = 'mytestsource__bitbucket__';

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig(mockttp.url);
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  test('process records from all streams', async () => {
    const cli = await CLI.runWith([
      'write',
      '--config',
      configPath,
      '--catalog',
      catalogPath,
      '--dry-run',
    ]);
    cli.stdin.end(bitbucketAllStreamsLog, 'utf8');

    const stdout = await read(cli.stdout);
    logger.debug(stdout);

    const processedByStream = {
      branches: 2,
      commits: 25,
      deployments: 14,
      issues: 1,
      pipeline_steps: 5,
      pipelines: 7,
      pull_request_activities: 12,
      pull_requests: 2,
      repositories: 6,
      workspace_users: 5,
      workspaces: 3,
    };
    const processed = _(processedByStream)
      .toPairs()
      .map((v) => [`${streamNamePrefix}${v[0]}`, v[1]])
      .orderBy(0, 'asc')
      .fromPairs()
      .value();

    const writtenByModel = {
      cicd_Build: 7,
      cicd_BuildCommitAssociation: 7,
      cicd_Deployment: 14,
      cicd_Organization: 3,
      cicd_Repository: 6,
      compute_Application: 7,
      tms_Project: 6,
      tms_Task: 1,
      tms_TaskAssignment: 1,
      tms_TaskBoard: 6,
      tms_TaskBoardProjectRelationship: 6,
      tms_TaskBoardRelationship: 1,
      tms_TaskProjectRelationship: 1,
      tms_User: 2,
      vcs_Branch: 2,
      vcs_Commit: 25,
      vcs_Membership: 5,
      vcs_Organization: 3,
      vcs_PullRequest: 2,
      vcs_PullRequestComment: 2,
      vcs_PullRequestReview: 3,
      vcs_Repository: 6,
      vcs_User: 33,
    };

    const processedTotal = _(processedByStream).values().sum();
    const writtenTotal = _(writtenByModel).values().sum();
    expect(stdout).toMatch(`Processed ${processedTotal} records`);
    expect(stdout).toMatch(`Would write ${writtenTotal} records`);
    expect(stdout).toMatch('Errored 0 records');
    expect(stdout).toMatch('Skipped 0 records');
    expect(stdout).toMatch(
      JSON.stringify(
        AirbyteLog.make(
          AirbyteLogLevel.INFO,
          `Processed records by stream: ${JSON.stringify(processed)}`
        )
      )
    );
    expect(stdout).toMatch(
      JSON.stringify(
        AirbyteLog.make(
          AirbyteLogLevel.INFO,
          `Would write records by model: ${JSON.stringify(writtenByModel)}`
        )
      )
    );
    expect(await read(cli.stderr)).toBe('');
    expect(await cli.wait()).toBe(0);
  });
});
