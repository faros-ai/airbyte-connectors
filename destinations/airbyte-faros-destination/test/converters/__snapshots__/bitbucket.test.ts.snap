// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bitbucket basic test process records from all streams 1`] = `
Array [
  "Processed 82 records",
  "Processed records by stream: {\\\\\\"mytestsource__bitbucket__branches\\\\\\":2,\\\\\\"mytestsource__bitbucket__commits\\\\\\":25,\\\\\\"mytestsource__bitbucket__deployments\\\\\\":14,\\\\\\"mytestsource__bitbucket__issues\\\\\\":1,\\\\\\"mytestsource__bitbucket__pipeline_steps\\\\\\":5,\\\\\\"mytestsource__bitbucket__pipelines\\\\\\":7,\\\\\\"mytestsource__bitbucket__pull_request_activities\\\\\\":12,\\\\\\"mytestsource__bitbucket__pull_requests\\\\\\":2,\\\\\\"mytestsource__bitbucket__repositories\\\\\\":6,\\\\\\"mytestsource__bitbucket__workspace_users\\\\\\":5,\\\\\\"mytestsource__bitbucket__workspaces\\\\\\":3}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 149 records",
  "Would write records by model: {\\\\\\"cicd_Build\\\\\\":7,\\\\\\"cicd_BuildCommitAssociation\\\\\\":7,\\\\\\"cicd_Deployment\\\\\\":14,\\\\\\"cicd_Organization\\\\\\":3,\\\\\\"cicd_Repository\\\\\\":6,\\\\\\"compute_Application\\\\\\":7,\\\\\\"tms_Project\\\\\\":6,\\\\\\"tms_Task\\\\\\":1,\\\\\\"tms_TaskAssignment\\\\\\":1,\\\\\\"tms_TaskBoard\\\\\\":6,\\\\\\"tms_TaskBoardProjectRelationship\\\\\\":6,\\\\\\"tms_TaskBoardRelationship\\\\\\":1,\\\\\\"tms_TaskProjectRelationship\\\\\\":1,\\\\\\"tms_User\\\\\\":2,\\\\\\"vcs_Branch\\\\\\":2,\\\\\\"vcs_Commit\\\\\\":25,\\\\\\"vcs_Membership\\\\\\":5,\\\\\\"vcs_Organization\\\\\\":3,\\\\\\"vcs_PullRequest\\\\\\":2,\\\\\\"vcs_PullRequestComment\\\\\\":2,\\\\\\"vcs_PullRequestReview\\\\\\":3,\\\\\\"vcs_Repository\\\\\\":6,\\\\\\"vcs_User\\\\\\":33}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;
